paymentsheet/src/main/java/com/stripe/android/common/ui/BottomSheet.kt

```diff
@@ -27,24 +27,39 @@ import androidx.compose.runtime.snapshotFlow
 import androidx.compose.ui.Modifier
 import androidx.compose.ui.graphics.Color
 import com.google.accompanist.systemuicontroller.rememberSystemUiController
+import kotlinx.coroutines.CancellationException
+import kotlinx.coroutines.CompletableDeferred
 import kotlinx.coroutines.flow.first
 
 @OptIn(ExperimentalMaterialApi::class)
 internal class BottomSheetState(
     val modalBottomSheetState: ModalBottomSheetState,
+    val keyboardHandler: BottomSheetKeyboardHandler,
 ) {
 
+    private var dismissalType: DismissalType? = null
+
     suspend fun show() {
-        modalBottomSheetState.show()
+        repeatUntilSucceeded {
+            modalBottomSheetState.show()
+        }
     }
 
-    suspend fun awaitDismissal() {
+    suspend fun awaitDismissal(): DismissalType {
         snapshotFlow { modalBottomSheetState.isVisible }.first { isVisible -> !isVisible }
+        return dismissalType ?: DismissalType.SwipedDownByUser
     }
 
     suspend fun hide() {
+        dismissalType = DismissalType.Programmatically
+        keyboardHandler.dismissAndWait()
```

Comment:
We dismiss the keyboard before we dismiss the sheet. This looks cleaner and prevents the `CancellationException`.

Reply:
Done!